pkg/distributor/distributor.go

```diff
@@ -423,10 +423,14 @@ func (d *Distributor) Push(ctx context.Context, req *client.WriteRequest) (*clie
 	}
 
 	now := time.Now()
-	if !d.ingestionRateLimiter.AllowN(now, userID, validatedSamples) {
+	if err := d.ingestionRateLimiter.WaitN(ctx, now, userID, validatedSamples); err != nil {
 		// Ensure the request slice is reused if the request is rate limited.
 		client.ReuseSlice(req.Timeseries)
 
+		if ctxErr := ctx.Err(); ctxErr != nil { // bail out if caller cancelled
+			return nil, ctxErr
+		}
+
```

Comment:
wouldn't `WaitN` have returned error already in such case?

Reply:
The point of this bit is not to count a cancellation by the caller as a discard.